Tests/AuthenticationInterceptorTests.swift

```diff
@@ -0,0 +1,663 @@
+
+        XCTAssertEqual(response?.result.isFailure, true)
+        XCTAssertEqual(response?.result.failure?.asAFError?.isResponseValidationError, true)
+        XCTAssertEqual(response?.result.failure?.asAFError?.responseCode, 500)
+
+        XCTAssertEqual(authenticator.applyCount, 1)
+        XCTAssertEqual(authenticator.refreshCount, 0)
+        XCTAssertEqual(authenticator.didRequestFailDueToAuthErrorCount, 1)
+        XCTAssertEqual(authenticator.isRequestAuthenticatedWithCredentialCount, 0)
+
+        XCTAssertEqual(request.retryCount, 0)
+    }
+
+    func testThatInterceptorThrowsMissingCredentialErrorWhenCredentialIsNilAndRequestShouldBeRetried() {
+        // Given
+        let credential = OAuthCredential()
+        let authenticator = OAuthAuthenticator()
+        let interceptor = AuthenticationInterceptor(authenticator: authenticator, credential: credential)
+
+        let eventMonitor = ClosureEventMonitor()
+        eventMonitor.requestDidCreateTask = { _, _ in interceptor.credential = nil }
+
+        let session = Session(eventMonitors: [eventMonitor])
+
+        let urlRequest = URLRequest.makeHTTPBinRequest(path: "status/401")
+
+        let expect = expectation(description: "request should complete")
+        var response: AFDataResponse<Data?>?
+
+        // When
+        let request = session.request(urlRequest, interceptor: interceptor).validate().response {
+            response = $0
+            expect.fulfill()
+        }
+
+        waitForExpectations(timeout: timeout, handler: nil)
+
+        // Then
+        XCTAssertEqual(response?.request?.value(forHTTPHeaderField: "Authorization"), "Bearer a0")
+
+        XCTAssertEqual(response?.result.isFailure, true)
+        XCTAssertEqual(response?.result.failure?.asAFError?.isRequestRetryError, true)
+        XCTAssertEqual(response?.result.failure?.asAFError?.underlyingError as? AuthenticationError, .missingCredential)
+
+        if case .requestRetryFailed(_, let originalError) = response?.result.failure {
+            XCTAssertEqual(originalError.asAFError?.isResponseValidationError, true)
+            XCTAssertEqual(originalError.asAFError?.responseCode, 401)
+        }
+
+        XCTAssertEqual(authenticator.applyCount, 1)
+        XCTAssertEqual(authenticator.refreshCount, 0)
+        XCTAssertEqual(authenticator.didRequestFailDueToAuthErrorCount, 1)
+        XCTAssertEqual(authenticator.isRequestAuthenticatedWithCredentialCount, 0)
+
+        XCTAssertEqual(request.retryCount, 0)
+    }
+
+    func testThatInterceptorRetriesRequestThatFailedWithOutdatedCredential() {
+        // Given
+        let credential = OAuthCredential()
+        let authenticator = OAuthAuthenticator()
+        let interceptor = AuthenticationInterceptor(authenticator: authenticator, credential: credential)
+
+        let eventMonitor = ClosureEventMonitor()
+
+        eventMonitor.requestDidCreateTask = { _, _ in
```

Comment:
I don't think there's really any other way to test this edge case. There's not actually any refresh that occurs here. This is testing the edge case where two requests are started in parallel, the first one fails with a 401, triggers refresh, and is retried with the new credential before the second request returns a 401. When the second request comes back with a 401, we want to avoid triggering a second refresh since we've already refreshed the original credential. As such, the second request should be immediately retried with the new credential rather than triggering a second refresh.

This test verifies the immediate retry logic that it does not trigger a second refresh. The best way I could find to actually make this work was to use a single request and swap the credential after the task is created. I think the chances of the request completing before the token can be swapped are really low since it has to go out to the network, but I do agree it is certainly racing.

The only way other way to really recreate this use case would be to try and delay a 401 request with httpbin, but that's not supported. It would also be even more racy than this approach. Of all the times I've run this test locally, I haven't seen any failures. It's something we could look to disable on CI if we ever see failure issues with it, but given that it has to round trip across the network faster than hopping onto the main queue to call the event handler, I don't think we're gonna fail there realistically. Even on super fast networks.

Reply:
Sounds right to me. Once I have Firewalk working for all tests, we'll see what happens with a local request.